roborock/cloud_api.py

```diff
@@ -112,40 +112,51 @@ def sync_start_loop(self) -> None:
             self._logger.info("Starting mqtt loop")
             super().loop_start()
 
-    def sync_disconnect(self) -> bool:
-        rc = mqtt.MQTT_ERR_AGAIN
+    def sync_disconnect(self) -> tuple[bool, Task[tuple[Any, VacuumError|None]] | None]:
+        if not self.is_connected():
+            return False, None
+
+        self._logger.info("Disconnecting from mqtt")
+        disconnected_future = asyncio.ensure_future(self._async_response(DISCONNECT_REQUEST_ID))
+        rc = super().disconnect()
+
+        if rc != mqtt.MQTT_ERR_SUCCESS:
+            disconnected_future.cancel()
+            disconnected_future = None
+
+        if rc not in [mqtt.MQTT_ERR_SUCCESS, mqtt.MQTT_ERR_NO_CONN]:
+            raise RoborockException(f"Failed to disconnect ({mqtt.error_string(rc)})")
+
+        return rc == mqtt.MQTT_ERR_SUCCESS, disconnected_future
+
+    def sync_connect(self) -> tuple[bool, Task[tuple[Any, VacuumError|None]] | None]:
         if self.is_connected():
-            self._logger.info("Disconnecting from mqtt")
-            rc = super().disconnect()
-            if rc not in [mqtt.MQTT_ERR_SUCCESS, mqtt.MQTT_ERR_NO_CONN]:
-                raise RoborockException(f"Failed to disconnect ({mqtt.error_string(rc)})")
-        return rc == mqtt.MQTT_ERR_SUCCESS
-
-    def sync_connect(self) -> bool:
-        should_connect = not self.is_connected()
-        if should_connect:
-            if self._mqtt_port is None or self._mqtt_host is None:
-                raise RoborockException("Mqtt information was not entered. Cannot connect.")
-            self._logger.info("Connecting to mqtt")
-            super().connect(host=self._mqtt_host, port=self._mqtt_port, keepalive=KEEPALIVE)
+            self.sync_start_loop()
+            return False, None
+
+        if self._mqtt_port is None or self._mqtt_host is None:
+            raise RoborockException("Mqtt information was not entered. Cannot connect.")
+
+        self._logger.info("Connecting to mqtt")
+        connected_future = asyncio.ensure_future(self._async_response(CONNECT_REQUEST_ID))
+        super().connect(host=self._mqtt_host, port=self._mqtt_port, keepalive=KEEPALIVE)
+
```

Comment:
Would be nice to add a secondary log for successfully connected

Reply:
There are logs in `on_connect` that should be enough

Reply:
Ah thanks. On mobile so I couldn’t see